### PR TITLE
Feature add passwordless email magic link

### PR DESCRIPTION
**Pull request by nikolaseu**

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -61,8 +61,17 @@
...
-            android:launchMode="singleTask"/>
-        <!--Auth0 Lock SMS End-->
+            android:launchMode="singleTask">
+            <intent-filter android:autoVerify="true">
+                <action android:name="android.intent.action.VIEW"/>
+                <category android:name="android.intent.category.DEFAULT"/>
+                <category android:name="android.intent.category.BROWSABLE"/>
+                <data android:scheme="https" />
+                <data android:scheme="http" />
+                <data android:host="nikolaseu-test.auth0.com" android:pathPrefix="/android/com.auth0.app.Lock/email"/>
```
Comment (hzalaz) [inline app/src/main/AndroidManifest.xml]: Change to `overmind.auth0.com`

File: lock-email/src/main/java/com/auth0/lock/email/LockEmailActivity.java [hunk 2]
```diff
@@ -151,10 +224,49 @@ protected void onStop() {
         finish();
     }
 
+    @SuppressWarnings("unused")
+    @Subscribe public void onLoginRequest(LoginRequestEvent event) {
+        email = event.getEmail();
+        passcode = event.getPasscode();
```
Comment (hzalaz) [inline lock-email/src/main/java/com/auth0/lock/email/LockEmailActivity.java]: Make `performLogin` receive the event itself and avoid storing the passcode.

```java
performLogin(event);
```

File: lock-email/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -36,13 +38,26 @@
...
+    <![CDATA[
+        Trying to log in with your email <b>%s</b>.
+    ]]>
+    </string>
     <string name="com_auth0_email_login_message">
     <![CDATA[
         Please check your email <b>%s</b>.<br/>You&#39;ve received a message from us with your passcode
     ]]>
     </string>
+    <string name="com_auth0_email_invalid_link_error_title">This link isn\'t valid anymore</string>
```
Comment (hzalaz) [inline lock-email/src/main/res/values/strings.xml]: `Couldn't log in with the magic link, it was probably used already or not valid anymore. Please request a new one`

**Comment by hzalaz** [comment]
@nikolaseu also add a screenshot of each new screen so I can take a look at them

**Comment by hzalaz** [comment]
When the link is not valid for whatever reason, add a button to quickly start the process again.

**Comment by nikolaseu** [comment]
screenshot-2015-11-13_17 04 08 523
screenshot-2015-11-13_17 04 25 346
screenshot-2015-11-13_17 04 55 883
screenshot-2015-11-13_17 11 37 308

**Comment by nikolaseu** [comment]
The "trying to log in ..." with the circular progress bar is kind of ugly

**Comment by hzalaz** [comment]
* Change `Email Magic Link` for `Magic Link Sent` title
* Both progress indicator and message should be centered horizontally.
* In the invalid link screen, remove the `Please request a new one` seeing the button is enough to show what can he/she do.

**Comment by hzalaz** [comment]
Instead of `Trying....` use `Logging in with your email <email_address>`

**Comment by hzalaz** [comment]
Also the `Didn't receive the link` should be `Let me enter a code` all in CAPS

**Comment by nikolaseu** [comment]
TODO:
- add "resend" button to resend the email with the link. this button should be hidden initially and shown after 10 secs
- the 'let me enter a code' should go to the 'enter code' fragment to allow the user to enter the code manually when the link doesn't work
- add 'green circle with a tick' above the message, similar to the iOS version

File: identity-core/src/main/java/com/auth0/identity/web/LinkParser.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
+package com.auth0.identity.web;
+
+import android.content.Intent;
+import android.net.Uri;
+import android.util.Log;
+
+public class LinkParser {
+    private static String TAG = LinkParser.class.getName();
+
+    public static String getCodeFromAppLinkIntent(Intent intent) {
```
Comment (hzalaz) [inline identity-core/src/main/java/com/auth0/identity/web/LinkParser.java]: Make this an instance method

File: identity-core/src/main/java/com/auth0/identity/web/LinkParser.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+            if (code != null) {
+                return code;
+            }
+        }
+
+        Log.d(TAG, "Invalid app link intent: " + intent);
+        return null;
+    }
+
+    public static String getCodeFromAppLinkUri(String uriString) {
```
Comment (hzalaz) [inline identity-core/src/main/java/com/auth0/identity/web/LinkParser.java]: Make this an instance method

File: app/src/main/java/com/auth0/app/MyActivity.java [hunk 6]
```diff
@@ -58,7 +58,9 @@ public void onClick(View v) {
         emailButton.setOnClickListener(new View.OnClickListener() {
             @Override
             public void onClick(View v) {
-                startActivity(new Intent(MyActivity.this, LockEmailActivity.class));
+                Intent intent = new Intent(MyActivity.this, LockEmailActivity.class);
```
Comment (hzalaz) [inline app/src/main/java/com/auth0/app/MyActivity.java]: Lets add the following static method to `LockEmailActivity`:

```java
public static void showFrom(Activity activity, boolean useMagicLink) {
    Intent intent = new Intent(activity, LockEmailActivity.class);
    intent.putExtra(LockEmailActivity.USE_MAGIC_LINK, useMagicLink);
    activity.startActivity(intent);
}
```

Then we can just show it like this

```java
LockEmailActivity.showFrom(this, true);
```

**Comment by nikolaseu** [comment]
screenshot-2015-11-16_14 24 35 708

**Comment by nikolaseu** [comment]
screenshot-2015-11-16_14 31 22 556

File: lock-email/src/main/java/com/auth0/lock/email/LockEmailActivity.java [hunk 7]
```diff
@@ -111,14 +176,42 @@ protected void onStop() {
...
+        if (!useMagicLink) {
+            fragment = new EmailLoginFragment();
+            arguments.putString(EmailLoginFragment.EMAIL_ARGUMENT, email);
+        } else {
+            fragment = new MagicLinkLoginFragment();
+            arguments.putString(MagicLinkLoginFragment.EMAIL_ARGUMENT, email);
+        }
+        fragment.setArguments(arguments);
+        // in case we were already showing the fragment (it happens we the user asks to resend the email)
+        getSupportFragmentManager().popBackStack(fragment.getClass().getName(), FragmentManager.POP_BACK_STACK_INCLUSIVE);
```
Comment (nikolaseu) [inline lock-email/src/main/java/com/auth0/lock/email/LockEmailActivity.java]: idk if this is the best way (but it's certainly the easiest)

**Pull request merged**